Add a "post rendering frame" hook

smfr opened the pull request:
#### bf82da4c0e3bbd409f44510fa9f3d1891808d770
<pre>
Add a &quot;post rendering frame&quot; hook
<a href="https://bugs.webkit.org/show_bug.cgi?id=249807">https://bugs.webkit.org/show_bug.cgi?id=249807</a>
rdar://103648840

Reviewed by Chris Dumez and Ryosuke Niwa.

We need to add a Page-level &quot;post rendering frame&quot; hook so we can do things like hook up
postRequestAnimationFrame() (webkit.org/b/249798), possibly run the WindowEventLoop
(webkit.org/b/249684) and correctly implement Inspectors &quot;frame end&quot; event (webkit.org/b/249796).

For Core Animation-driven drawing clients (TiledCoreAnimationDrawingArea, WebKitLegacy) we want this
post-rendering frame hook to run after CA&apos;s runloop observer returns, since this callback may run
arbitrary script at some point and we need to avoid any possible re-entrancy issues. So we use a
second CFRunLoopObserver, ordered as CoreAnimationCommit+2 (after InspectorFrameEnd).

For RemoteLayerTreeDrawingArea, we can just call the entry point after
didCompleteRenderingUpdateDisplay().

* Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
(WebCore::InspectorTimelineAgent::internalStart):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didCompleteRenderingFrame):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/cf/RunLoopObserver.cpp:
(WebCore::RunLoopObserver::schedule):
* Source/WebCore/platform/cf/RunLoopObserver.h: Modernize terminology: LayerFlush -&gt; RenderingUpdate and add PostRenderingUpdate.
* Source/WebKit/WebProcess/WebPage/DrawingArea.cpp:
(WebKit::DrawingArea::didCompleteRenderingFrame):
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didCompleteRenderingFrame):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::TiledCoreAnimationDrawingArea):
(WebKit::TiledCoreAnimationDrawingArea::~TiledCoreAnimationDrawingArea):
(WebKit::TiledCoreAnimationDrawingArea::setLayerTreeStateIsFrozen):
(WebKit::TiledCoreAnimationDrawingArea::dispatchAfterEnsuringUpdatedScrollPosition):
(WebKit::TiledCoreAnimationDrawingArea::didCompleteRenderingUpdateDisplay):
(WebKit::TiledCoreAnimationDrawingArea::scheduleRenderingUpdateRunLoopObserver):
(WebKit::TiledCoreAnimationDrawingArea::invalidateRenderingUpdateRunLoopObserver):
(WebKit::TiledCoreAnimationDrawingArea::renderingUpdateRunLoopCallback):
(WebKit::TiledCoreAnimationDrawingArea::schedulePostRenderingUpdateRunLoopObserver):
(WebKit::TiledCoreAnimationDrawingArea::invalidatePostRenderingUpdateRunLoopObserver):
(WebKit::TiledCoreAnimationDrawingArea::postRenderingUpdateRunLoopCallback):
(WebKit::TiledCoreAnimationDrawingArea::updateRenderingRunLoopCallback): Deleted.
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _didCompleteRenderingUpdateDisplay]):
* Source/WebKitLegacy/mac/WebView/WebViewData.mm:
(WebViewLayerFlushScheduler::WebViewLayerFlushScheduler):

Canonical link: <a href="https://commits.webkit.org/258311@main">https://commits.webkit.org/258311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/118d365780a18b63524cf2eb368ce62ce8b8211b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101477 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10633 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110748 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171003 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105457 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11587 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1531 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93870 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108565 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107258 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8817 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92062 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35348 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90714 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23471 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78347 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4235 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24980 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4298 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1436 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10385 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44469 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5706 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6060 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
```diff
@@ -368,8 +368,10 @@
         RunLoop::main().dispatch([pageID] {
             if (auto* webPage = WebProcess::singleton().webPage(pageID)) {
-                if (auto* drawingArea = dynamicDowncast<RemoteLayerTreeDrawingArea>(webPage->drawingArea()))
+                if (auto* drawingArea = dynamicDowncast<RemoteLayerTreeDrawingArea>(webPage->drawingArea())) {
```
Can we add an assertion that this is back to the main/web thread?

cdumez commented on Source/WebKitLegacy/mac/WebView/WebViewData.mm:
```diff
@@ -140,6 +156,16 @@ static CFRunLoopRef currentRunLoop()
+void WebViewLayerFlushScheduler::postRenderingUpdateCallback()
+{
+    @autoreleasepool {
+        RefPtr<LayerFlushController> protector = m_flushController;
```
nit: We don't need to specify the whole type nowadays:
```
RefPtr protectedFlushController { m_flushController };
```

rniwa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1938,6 +1938,15 @@ void Page::didCompleteRenderingUpdateDisplay()
     m_inspectorController->didComposite(mainFrame());
 }
 
+void Page::didCompleteRenderingFrame()
```
It would be ideal to all these updating per `WindowEventLoop` object. e.g. cross-site contents in a single Page may not share EventLoop object (this is especially true once site isolation is implemented).

smfr replied:
Agreed, but I'd rather do that later. It should be pretty clean; the Page `updateRendering` code is in a good state.

cdumez reviewed (approved):
LGTM

smfr committed:
Add a "post rendering frame" hook
https://bugs.webkit.org/show_bug.cgi?id=249807
rdar://103648840

Reviewed by Chris Dumez and Ryosuke Niwa.

We need to add a Page-level "post rendering frame" hook so we can do things like hook up
postRequestAnimationFrame() (webkit.org/b/249798), possibly run the WindowEventLoop
(webkit.org/b/249684) and correctly implement Inspectors "frame end" event (webkit.org/b/249796).

For Core Animation-driven drawing clients (TiledCoreAnimationDrawingArea, WebKitLegacy) we want this
post-rendering frame hook to run after CA's runloop observer returns, since this callback may run
arbitrary script at some point and we need to avoid any possible re-entrancy issues. So we use a
second CFRunLoopObserver, ordered as CoreAnimationCommit+2 (after InspectorFrameEnd).

For RemoteLayerTreeDrawingArea, we can just call the entry point after
didCompleteRenderingUpdateDisplay().

* Source/WebCore/inspector/agents/InspectorTimelineAgent.cpp:
(WebCore::InspectorTimelineAgent::internalStart):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didCompleteRenderingFrame):
* Source/WebCore/page/Page.h:
* Source/WebCore/platform/cf/RunLoopObserver.cpp:
(WebCore::RunLoopObserver::schedule):
* Source/WebCore/platform/cf/RunLoopObserver.h: Modernize terminology: LayerFlush -> RenderingUpdate and add PostRenderingUpdate.
* Source/WebKit/WebProcess/WebPage/DrawingArea.cpp:
(WebKit::DrawingArea::didCompleteRenderingFrame):
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateRendering):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::didCompleteRenderingFrame):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/mac/TiledCoreAnimationDrawingArea.mm:
(WebKit::TiledCoreAnimationDrawingArea::TiledCoreAnimationDrawingArea):
(WebKit::TiledCoreAnimationDrawingArea::~TiledCoreAnimationDrawingArea):
(WebKit::TiledCoreAnimationDrawingArea::setLayerTreeStateIsFrozen):
(WebKit::TiledCoreAnimationDrawingArea::dispatchAfterEnsuringUpdatedScrollPosition):
(WebKit::TiledCoreAnimationDrawingArea::didCompleteRenderingUpdateDisplay):
(WebKit::TiledCoreAnimationDrawingArea::scheduleRenderingUpdateRunLoopObserver):
(WebKit::TiledCoreAnimationDrawingArea::invalidateRenderingUpdateRunLoopObserver):
(WebKit::TiledCoreAnimationDrawingArea::renderingUpdateRunLoopCallback):
(WebKit::TiledCoreAnimationDrawingArea::schedulePostRenderingUpdateRunLoopObserver):
(WebKit::TiledCoreAnimationDrawingArea::invalidatePostRenderingUpdateRunLoopObserver):
(WebKit::TiledCoreAnimationDrawingArea::postRenderingUpdateRunLoopCallback):
(WebKit::TiledCoreAnimationDrawingArea::updateRenderingRunLoopCallback): Deleted.
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _didCompleteRenderingUpdateDisplay]):
* Source/WebKitLegacy/mac/WebView/WebViewData.mm:
(WebViewLayerFlushScheduler::WebViewLayerFlushScheduler):

Canonical link: https://commits.webkit.org/258311@main
